File: python/q2_train_frozen_model.py
from __future__ import annotations

import argparse
from datetime import datetime, timezone
from pathlib import Path
import joblib

from common import MODEL_ORDER, candidate_feature_counts, fit_frozen_model, load_xy, package_versions, write_json


def parse_k_values(text):
    if not text:
        return None
    return [int(x.strip()) for x in text.split(",") if x.strip()]


def read_model_name(path, explicit):
    if explicit:
        return explicit
    if path:
        return Path(path).read_text(encoding="utf-8").strip()
    raise ValueError("Provide either --model or --recommended-model-file.")


def main():
    parser = argparse.ArgumentParser(description="Q2: fit and serialize the final transfer-ready frozen classifier after model choice is locked.")
    parser.add_argument("--data", required=True)
    parser.add_argument("--label", required=True)
    parser.add_argument("--id-column", default=None)
    parser.add_argument("--positive", default=None)
    parser.add_argument("--model", default=None, choices=MODEL_ORDER)
    parser.add_argument("--recommended-model-file", default=None)
    parser.add_argument("--feature-counts", default=None)
    parser.add_argument("--inner-folds", type=int, default=5)
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--output", default="artifacts/python_frozen_model")
    args = parser.parse_args()

    out = Path(args.output)
    out.mkdir(parents=True, exist_ok=True)
    X, y, labels, _ = load_xy(args.data, args.label, args.id_column, args.positive)
    k_values = candidate_feature_counts(X.shape[1], parse_k_values(args.feature_counts))
    model_name = read_model_name(args.recommended_model_file, args.model)
    if model_name not in MODEL_ORDER:
        raise ValueError(f"Unsupported model '{model_name}'. Supported: {MODEL_ORDER}")

    final_pipe, selected_features, best_params, inner_auc = fit_frozen_model(X, y, model_name, k_values, args.inner_folds, args.seed)
    artifact = {
        "pipeline": final_pipe,
        "selected_features": selected_features,
        "model_name": model_name,
        "positive_label": labels.positive,
        "negative_label": labels.negative,
        "training_feature_columns": X.columns.tolist(),
        "best_search_params": best_params,
        "inner_cv_roc_auc": inner_auc,
        "seed": args.seed,
        "package_versions": package_versions(),
    }
    joblib.dump(artifact, out / "frozen_model.joblib")
    (out / "selected_features.txt").write_text("\n".join(selected_features) + "\n", encoding="utf-8")
    metadata = {k: v for k, v in artifact.items() if k != "pipeline"}
    metadata.update({
        "created_utc": datetime.now(timezone.utc).isoformat(),
        "question": "Q2 - train a frozen model for transfer",
        "important": "Performance should be reported from Q3 nested CV and/or an untouched external validation cohort, not from this full-development-data refit.",
    })
    write_json(out / "frozen_model_metadata.json", metadata)


if __name__ == "__main__":
    main()

File: R/q2_train_frozen_model.R
full_args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", full_args, value = TRUE)
script_dir <- if (length(file_arg)) dirname(normalizePath(sub("^--file=", "", file_arg[[1]]))) else getwd()
source(file.path(script_dir, "common.R"))

args <- parse_cli_args()
data_path <- arg_value(args, "data", required = TRUE)
label_col <- arg_value(args, "label", required = TRUE)
id_col <- arg_value(args, "id-column", default = NULL)
positive <- arg_value(args, "positive", default = NULL)
out_dir <- arg_value(args, "output", default = "artifacts/R_frozen_model")
inner_folds <- as.integer(arg_value(args, "inner-folds", 5))
seed <- as.integer(arg_value(args, "seed", 42))
model_name <- arg_value(args, "model", default = NULL)
recommended_file <- arg_value(args, "recommended-model-file", default = NULL)
if (is.null(model_name) && !is.null(recommended_file)) model_name <- trimws(readLines(recommended_file, warn = FALSE)[[1]])
if (is.null(model_name)) stop("Provide either --model or --recommended-model-file.")
if (!model_name %in% MODEL_ORDER) stop("Unsupported model: ", model_name)

input <- load_xy(data_path, label_col, id_col, positive)
k_values <- candidate_feature_counts(ncol(input$x), parse_int_list(arg_value(args, "feature-counts", NULL)))
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)

artifact <- fit_frozen_model(input$x, input$y, model_name, k_values, inner_folds, seed)
artifact$positive_label <- input$positive
artifact$negative_label <- input$negative
artifact$training_feature_columns <- colnames(input$x)
artifact$seed <- seed
artifact$package_versions <- list(
  R = R.version.string,
  caret = as.character(utils::packageVersion("caret")),
  glmnet = as.character(utils::packageVersion("glmnet")),
  randomForest = as.character(utils::packageVersion("randomForest")),
  e1071 = as.character(utils::packageVersion("e1071")),
  pROC = as.character(utils::packageVersion("pROC"))
)
saveRDS(artifact, file.path(out_dir, "frozen_model.rds"))
writeLines(artifact$selected_features, file.path(out_dir, "selected_features.txt"))
write_json(file.path(out_dir, "frozen_model_metadata.json"), list(
  question = "Q2 - train a frozen model for transfer",
  model_name = model_name,
  selected_features = artifact$selected_features,
  positive_label = input$positive,
  negative_label = input$negative,
  best_search_params = artifact$best_search_params,
  inner_cv_roc_auc = artifact$inner_cv_roc_auc,
  seed = seed,
  package_versions = artifact$package_versions,
  important = "Performance should be reported from Q3 nested CV and/or an untouched external validation cohort, not from this full-development-data refit."
))
